### PR TITLE
ARROW-9383: [Python] Support fsspec filesystems in Dataset API

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

File: python/pyarrow/dataset.py [hunk 1]
```diff
@@ -239,15 +240,18 @@ def _ensure_filesystem(fs_or_uri):
...
             filesystem = SubTreeFileSystem(prefix, filesystem)
         return filesystem, is_local
-    elif isinstance(fs_or_uri, (LocalFileSystem, _MockFileSystem)):
-        return fs_or_uri, True
-    elif isinstance(fs_or_uri, FileSystem):
-        return fs_or_uri, False
-    else:
+
+    try:
+        filesystem = _ensure_filesystem(fs_or_uri)
```
Comment (pitrou) [inline python/pyarrow/dataset.py:245]: It's a bit confusing to have those two functions named the same. Can you rename one or other to distinguish them better?

File: python/pyarrow/fs.py [hunk 2]
```diff
@@ -63,6 +63,31 @@ def __getattr__(name):
...
+    # handle fsspec-compatible filesystems
+    try:
+        import fsspec
+        if isinstance(filesystem, fsspec.AbstractFileSystem):
+            if type(filesystem).__name__ == 'LocalFileSystem':
+                # In case its a simple LocalFileSystem, use native arrow one
+                return LocalFileSystem()
+            return PyFileSystem(FSSpecHandler(filesystem))
+    except ImportError:
+        pass
```
Comment (pitrou) [inline python/pyarrow/fs.py:79]: For readability, I would prefer:
```python
try:
    import fsspec
except ImportError:
    pass
else:
    ...
```

File: python/pyarrow/fs.py [hunk 3]
```diff
@@ -63,6 +63,31 @@ def __getattr__(name):
...
+
+    from pyarrow.filesystem import LocalFileSystem as LegacyLocalFileSystem
+
+    # map old filesystems to new ones
+    if isinstance(filesystem, LegacyLocalFileSystem):
+        return LocalFileSystem()
+    # TODO handle HDFS?
+
+    raise TypeError("Unrecognized fielsystem: {}".format(type(filesystem)))
```
Comment (pitrou) [inline python/pyarrow/fs.py:88]: typo: "filesystem"

**Review by pitrou — COMMENTED** [review]
LGTM, just a couple style issues

File: python/pyarrow/dataset.py (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:245]: Renamed the one in dataset.py

File: python/pyarrow/fs.py (hunk 2 above)
Comment (jorisvandenbossche) [reply python/pyarrow/fs.py:79]: Ah, yes, that's better

File: python/pyarrow/fs.py [hunk 4]
```diff
@@ -63,6 +63,32 @@ def __getattr__(name):
...
+                # In case its a simple LocalFileSystem, use native arrow one
+                return LocalFileSystem()
+            return PyFileSystem(FSSpecHandler(filesystem))
+
+    # map old filesystems to new ones
+    from pyarrow.filesystem import LocalFileSystem as LegacyLocalFileSystem
+
+    if isinstance(filesystem, LegacyLocalFileSystem):
+        return LocalFileSystem()
+    # TODO handle HDFS?
```
Comment (jorisvandenbossche) [inline python/pyarrow/fs.py:87]: Do you have any idea about this one? The "legacy" one defined in `io-hdfs.pxi` is backed by the C++ `arrow::io::HadoopFileSystem`, and the new filesystem is backed by `arrow::fs::HadoopFileSystem` which itself holds a `arrow::io::HadoopFileSystem` as "client". 
So naively I would say (with some additional methods), one should be able to make a new HadoopFileSystem by wrapping the C++ "client" of the legacy filesystem?

Comment (pitrou) [reply python/pyarrow/fs.py:87]: Perhaps, but it's also not very important. Let's not bother about it until there's a real need for it.

**Comment by jorisvandenbossche** [comment]
@rjzamora latest master now also supports fsspec filesystems in the high-level dataset API, so no need to do the fsspec -> pyarrow.fs filesystem conversion on the dask side.

**Pull request closed without merging**